### PR TITLE
Add tokemak subgraph

### PR DESCRIPTION
**Pull request by DhruvJain1122**
Tokemak subgraph is based on yield schema

### CODE REVIEW

**Comment by this-username-is-taken** [comment]
Pool and vault ABIs are identical? It seems like they only have a pool contract and not a vault contract. If that's the case, you should just have the pool ABI instead, and map that to the vault entity

File: subgraphs/tokemak/subgraph.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        - name: Oracle
+          file: ./abis/Oracle.json
+        - name: CalculationsCurve
+          file: ./abis/CalculationsCurve.json
+        - name: CalculationsSushiSwap
+          file: ./abis/CalculationsSushiSwap.json 
+      callHandlers:
+        - function: deposit(uint256)
+          handler: handleDeposit
+        - function: deposit(uint256,address)
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/subgraph.yaml:57]: Is this right? I only see a `depositFor` in their contract: https://github.com/Tokemak/tokemak-smart-contracts-public/blob/d4687c5fd1c852852a8dac90ee627ac6bdb74bc0/contracts/pools/Pool.sol#L87

File: subgraphs/tokemak/subgraph.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      abis:
+        - name: Manager
+          file: ./abis/Manager.json
+        - name: Vault
+          file: ./abis/Vault.json
+        - name: ERC20
+          file: ./abis/ERC20Detailed.json
+      eventHandlers:
+        - event: PoolRegistered(address)
+          handler: handleNewVault
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/subgraph.yaml:27]: this should be called `handlePoolRegistered`

File: subgraphs/tokemak/src/mappings/vaultMappings.ts [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+
+export function handleWithdraw(call: WithdrawCall): void {
+  log.info('[Vault mappings] Handle withdraw with shares. TX hash: {}', [
+    call.transaction.hash.toHexString(),
+  ]);
+  const vaultAddress = call.to;
+  let vault = VaultStore.load(vaultAddress.toHexString());
+  if (vault) {
+    const sharesBurnt    = call.inputs.requestedAmount
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/src/mappings/vaultMappings.ts:75]: `const requestedAmount = call.inputs.requestedAmount`

File: subgraphs/tokemak/subgraph.yaml (hunk 1 above)
Comment (DhruvJain1122) [reply subgraphs/tokemak/subgraph.yaml:57]: https://github.com/Tokemak/tokemak-smart-contracts-public/blob/d4687c5fd1c852852a8dac90ee627ac6bdb74bc0/contracts/pools/Pool.sol#L83

Comment (DhruvJain1122) [reply subgraphs/tokemak/subgraph.yaml]: Done

File: subgraphs/tokemak/src/mappings/vaultMappings.ts (hunk 3 above)
Comment (DhruvJain1122) [reply subgraphs/tokemak/src/mappings/vaultMappings.ts]: Done

File: subgraphs/tokemak/subgraph.yaml (hunk 2 above)
Comment (DhruvJain1122) [reply subgraphs/tokemak/subgraph.yaml]: Done

File: subgraphs/tokemak/src/mappings/managerMappings.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    protocol = new YieldAggregator(PROTOCOL_ID)
+    protocol.name = "Tokemak"
+    protocol.slug = "tokemak"
+    protocol.network = Network.ETHEREUM
+    protocol.type = ProtocolType.YIELD
+    protocol.vaultIds = []
+    protocol.save()
+  }
+}
+function createRewardTokens(): void{
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/src/mappings/managerMappings.ts:39]: new line in between functions

File: subgraphs/tokemak/src/mappings/managerMappings.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  const address = Address.fromString(TOKE_ADDRESS)
+  getOrCreateRewardToken(address)
+}
+export function handlePoolRegistered(event: PoolRegistered): void {
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/src/mappings/managerMappings.ts:43]: https://github.com/messari/subgraphs/blob/master/docs/Mapping.md#assemblyscript-issues

File: subgraphs/tokemak/src/mappings/managerMappings.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  createProtocol()
+  createRewardTokens()
+  getOrCreateVault(event.params.pool, event)
+}
+
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/src/mappings/managerMappings.ts:48]: just one line in between functions

File: subgraphs/tokemak/src/mappings/managerMappings.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    const rewardToken = getOrCreateRewardToken(Address.fromString(TOKE_ADDRESS))
+
+    vault.rewardTokens = [rewardToken.id]
+
+    const managementFeeId = "management-fee-" + vault.id
+    let managementFee = new VaultFee(managementFeeId)
+    managementFee.feeType = VaultFeeType.MANAGEMENT_FEE
+    //TODOAM
+    managementFee.feePercentage =BIGDECIMAL_ZERO
+    // managementFee.feePercentage = 0
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/src/mappings/managerMappings.ts:80]: remove unused code

File: subgraphs/tokemak/src/mappings/managerMappings.ts [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+    const rewardToken = getOrCreateRewardToken(Address.fromString(TOKE_ADDRESS))
+
+    vault.rewardTokens = [rewardToken.id]
+
+    const managementFeeId = "management-fee-" + vault.id
+    let managementFee = new VaultFee(managementFeeId)
+    managementFee.feeType = VaultFeeType.MANAGEMENT_FEE
+    //TODOAM
+    managementFee.feePercentage =BIGDECIMAL_ZERO
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/src/mappings/managerMappings.ts:79]: spacing

File: subgraphs/tokemak/src/mappings/managerMappings.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    vault.createdTimestamp = event.block.timestamp
+
+    const rewardToken = getOrCreateRewardToken(Address.fromString(TOKE_ADDRESS))
+
+    vault.rewardTokens = [rewardToken.id]
+
+    const managementFeeId = "management-fee-" + vault.id
+    let managementFee = new VaultFee(managementFeeId)
+    managementFee.feeType = VaultFeeType.MANAGEMENT_FEE
+    //TODOAM
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/src/mappings/managerMappings.ts:78]: what's this?

File: subgraphs/tokemak/src/mappings/managerMappings.ts [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+    const performanceFeeId = "performance-fee-" + vault.id
+    let performanceFee = new VaultFee(performanceFeeId)
+    performanceFee.feeType = VaultFeeType.PERFORMANCE_FEE
+    //TODOAM
+    performanceFee.feePercentage =BIGDECIMAL_ZERO
+    // performanceFee.feePercentage = 0
+    performanceFee.save()
+
+    vault.fees = [managementFeeId, performanceFeeId]
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/src/mappings/managerMappings.ts:91]: `vault.fees = []` since no fee

**Comment by this-username-is-taken** [comment]
You are missing actions for usage metrics. See https://github.com/messari/subgraphs/blob/master/docs/Schema.md#usage-metrics

File: subgraphs/tokemak/src/mappings/managerMappings.ts [hunk 11]
```diff
@@ -75,25 +76,23 @@ function getOrCreateVault(vaultAddress: Address, event: ethereum.Event): VaultSt
     const managementFeeId = "management-fee-" + vault.id
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/src/mappings/managerMappings.ts:76]: I guess we don't need this anymore

File: subgraphs/tokemak/src/mappings/managerMappings.ts [hunk 12]
```diff
@@ -75,25 +76,23 @@ function getOrCreateVault(vaultAddress: Address, event: ethereum.Event): VaultSt
...
     let managementFee = new VaultFee(managementFeeId)
     managementFee.feeType = VaultFeeType.MANAGEMENT_FEE
-    //TODOAM
-    managementFee.feePercentage =BIGDECIMAL_ZERO
-    // managementFee.feePercentage = 0
+    managementFee.feePercentage = BIGDECIMAL_ZERO
     managementFee.save()
 
     const performanceFeeId = "performance-fee-" + vault.id
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/src/mappings/managerMappings.ts:82]: same here

File: subgraphs/tokemak/src/common/constants.ts [hunk 13]
```diff
@@ -1,84 +1,77 @@
...
 
 export namespace VaultFeeType {
-  export const MANAGEMENT_FEE = "MANAGEMENT_FEE"
-  export const PERFORMANCE_FEE = "PERFORMANCE_FEE"
-  export const DEPOSIT_FEE = "DEPOSIT_FEE"
-  export const WITHDRAWLAL_FEE = "WITHDRAWLAL_FEE"
+  export const MANAGEMENT_FEE = "MANAGEMENT_FEE";
+  export const PERFORMANCE_FEE = "PERFORMANCE_FEE";
+  export const DEPOSIT_FEE = "DEPOSIT_FEE";
+  export const WITHDRAWLAL_FEE = "WITHDRAWLAL_FEE";
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/src/common/constants.ts:35]: and this

File: subgraphs/tokemak/src/common/utils.ts [hunk 14]
```diff
@@ -1,8 +1,4 @@
-import {
-  BigInt,
-  BigDecimal,
-  ethereum,
-} from '@graphprotocol/graph-ts';
+import { BigInt, BigDecimal, ethereum } from "@graphprotocol/graph-ts";
 
 export function getTimeInMillis(time: BigInt): BigInt {
   return time.times(BigInt.fromI32(1000));
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/src/common/utils.ts:4]: use `BIGINT_THOUSAND` from constant.ts

File: subgraphs/tokemak/src/common/utils.ts [hunk 15]
```diff
@@ -13,5 +9,5 @@ export function getTimestampInMillis(block: ethereum.Block): BigInt {
 }
 
 export function bigIntToPercentage(n: BigInt): BigDecimal {
-  return n.toBigDecimal().div(BigDecimal.fromString("100"))
+  return n.toBigDecimal().div(BigDecimal.fromString("100"));
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/src/common/utils.ts:12]: can you add `BIGDECIMAL_HUNDRED` to `constant.ts` (also in reference subgraph) and use that here?

File: subgraphs/tokemak/src/common/tokens.ts [hunk 16]
```diff
@@ -26,17 +26,19 @@ export function getOrCreateToken(address: Address): Token {
 export function getOrCreateRewardToken(address: Address): RewardToken {
   let id = address.toHexString();
   let token = RewardToken.load(id);
-  if (!token) {
+  if (!token || token.name === "") {
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/src/common/tokens.ts:29]: why do we need `|| token.name === ""` if we are already hardcoding name and symbol?

Comment (DhruvJain1122) [reply subgraphs/tokemak/src/common/tokens.ts:29]: For new vaults

File: subgraphs/tokemak/src/common/tokens.ts [hunk 17]
```diff
@@ -26,17 +26,19 @@ export function getOrCreateToken(address: Address): Token {
...
     // Using try_cause some values might be missing
     let name = erc20Contract.try_name();
     let symbol = erc20Contract.try_symbol();
-    // TODO: add overrides for name and symbol
     token.decimals = decimals.reverted ? DEFAULT_DECIMALS : decimals.value;
-    token.name = name.reverted ? "" : name.value;
-    token.symbol = symbol.reverted ? "" : symbol.value;
+
+    //TOKE is the only reward token
+    token.name = name.reverted ? "Tokemak" : name.value;
```
Comment (this-username-is-taken) [inline subgraphs/tokemak/src/common/tokens.ts:39]: You can't do this. If we have another reward token then the logic would be wrong

Comment (this-username-is-taken) [reply subgraphs/tokemak/src/common/tokens.ts:39]: it should be
```
if (address == "TOKE_ADDRESS") {
  token.name = "Tokemak";
  token.symbol = "TOKE";
} else {
  ...
}
```

**Pull request merged**